### PR TITLE
Implement systems, domains and resources

### LINKED ISSUE #3765: catalog model: introduce the System, Domain, and Resource kinds
## Feature Suggestion

Before removing the alpha label on the catalog, I think it would be sensible to have at least the `System` and `Domain` kinds in place, and preferably also an outline of `Resource`, as defined in [the system model RFC](https://github.com/backstage/backstage/issues/390).

## Possible Implementation

I would argue that there is value in adding these in a very barebones manner, and iterate. What we need is:

- The skeleton of the kinds themselves
- Adding `system` and `domain` fields in the relevant `spec`s of various kinds where appropriate, such as `Component`
- Defining the relations that the `system` and `domain` fields translate to, probably new ones along the lines of `partOf`, since the [existing ones](https://backstage.io/docs/features/software-catalog/well-known-relations) don't seem like a great match.

## Context

See the following RFCs:
- [Backstage System Model](https://github.com/backstage/backstage/issues/390)
- [On specific semantic of systems, components, and resources in Catalog](https://github.com/backstage/backstage/issues/2898)

### PR DESCRIPTION
**Pull request by Fox32**
This PR implements the kinds System, Domain and Resource as defined by the previous PR #3957. This adds the entity definition, validation and relation emission, as well as some test data. I hope the test data makes somehow sense 😆 

Closes #3765

Needs to be rebased once #4031 is merged.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: app-config.yaml [hunk 1]
```diff
@@ -183,7 +183,16 @@ catalog:
...
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-resources.yaml
+    # Backstage example resources
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-systems.yaml
+    # Backstage example systems
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-domains.yaml
+    # Backstage example domains
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-apis.yaml
```
Comment (Fox32) [inline app-config.yaml:195]: I wonder why we don't use relative path here, that would make local development easier? Like:


```yaml
    - type: file
      target: ../catalog-model/examples/all-domains.yaml
```

File: packages/catalog-model/examples/domains/playback-domain.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: backstage.io/v1alpha1
+kind: Domain
+metadata:
+  name: playback
+  description: Everything related to audio playback
+spec:
+  owner: User/frank.tiernan
```
Comment (Fox32) [inline packages/catalog-model/examples/domains/playback-domain.yaml:7]: By intention a user, to test that case too

File: packages/catalog-model/src/kinds/SystemEntityV1alpha1.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    })
+    .required(),
+});
+
+export interface SystemEntityV1alpha1 extends Entity {
+  apiVersion: typeof API_VERSION[number];
+  kind: typeof KIND;
+  spec: {
+    owner: string;
+    domain?: string;
```
Comment (Fox32) [inline packages/catalog-model/src/kinds/SystemEntityV1alpha1.ts:40]: Maybe systems have a `lifecycle` too, but that is something for the future.

**Comment by freben** [comment]
@Fox32 now you've run into some conflicts :P it's moving ahead pretty fast, loving it

File: docs/features/software-catalog/descriptor-format.md [hunk 4]
```diff
@@ -383,8 +383,6 @@ spec:
   lifecycle: production
   owner: artist-relations-team
   system: artist-engagement-portal
-  providesApis:
```
Comment (tragiclifestories) [inline docs/features/software-catalog/descriptor-format.md:386]: Should this have disappeared?

File: packages/catalog-model/examples/domains/playback-domain.yaml (hunk 2 above)
Comment (tragiclifestories) [reply packages/catalog-model/examples/domains/playback-domain.yaml:7]: shouldnt't this be `user:frank.tiernan`? I thought the slash was for namespaces ...

**Review by tragiclifestories — COMMENTED** [review]
Couple of drive-by comments

File: packages/catalog-model/src/kinds/relations.ts [hunk 5]
```diff
@@ -57,8 +57,13 @@ export const RELATION_MEMBER_OF = 'memberOf';
 export const RELATION_HAS_MEMBER = 'hasMember';
 
 /**
+<<<<<<< HEAD
```
Comment (freben) [inline packages/catalog-model/src/kinds/relations.ts:60]: merge leftovers

File: plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts [hunk 6]
```diff
@@ -124,12 +144,13 @@ describe('BuiltinKindsEntityProcessor', () => {
           owner: 'o',
           lifecycle: 'l',
           definition: 'd',
+          system: 's',
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts:147]: we actually said to not do APIs in systems yet, right?

File: docs/features/software-catalog/descriptor-format.md (hunk 4 above)
Comment (Fox32) [reply docs/features/software-catalog/descriptor-format.md:386]: Hmm you are right. For some reason I wanted to remove this 🤔 😆 No it has to stay!

File: plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts (hunk 6 above)
Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts:147]: No we haven't. This says that an API can belong to a system. But system cannot expose or consume APIs

File: app-config.yaml (hunk 1 above)
Comment (Fox32) [reply app-config.yaml:195]: This already works, I use it in my `app-config.local.yaml` and we also use it in our app for testing.

Comment (freben) [reply app-config.yaml:195]: Hmmm. Sounds like a PR is coming up then! :) If people build and run in docker and other more esoteric situations though, it'll break.

File: plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts (hunk 6 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.test.ts:147]: Aha! Yeah my bad.

File: app-config.yaml [hunk 7]
```diff
@@ -184,6 +193,15 @@ catalog:
...
       target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-apis.yaml
+    # Backstage example resources
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-resources.yaml
+    # Backstage example systems
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-systems.yaml
+    # Backstage example domains
+    - type: url
+      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-domains.yamlp
```
Comment (freben) [inline app-config.yaml:204]:
```suggestion
      target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/all-domains.yaml
```

Comment (Fox32) [reply app-config.yaml]: Looks like I tried to press COMMAND+P 😆

File: app-config.yaml (hunk 1 above)
Comment (Fox32) [reply app-config.yaml]: @freben Btw, was that a request to created that PR? 😆 In that case I do it after this one to avoid conflicts.

**Review by freben — APPROVED** [review]
Awesome. This goes into the release today!

**Review by Rugvip — APPROVED** [review]

**Pull request merged**